### DISCUSSION TITLE
TSML UI - Update Meeting Info button opens customized form?

### DISCUSSION THREAD

**Discussion by Grateful-Viv** [code4recovery/12-step-meeting-list#823] [Q&A] [answered]
Is it possible to have the Update Meeting Info link located on the bottom left of a meeting open a form instead? I would do the form. It would just have to direct to a URL rather than trigger the default email.

**Comment by kiyote33** [comment]
Hi @Grateful-Viv,
Take a look at our 12 Step Meeting List Enhanced Feedback plugin, which replaces the standard feedback form. It only works with TSML but there is something in the works to have the same for TSML UI.

**Reply by Grateful-Viv** [reply]
I installed it, but I was hoping to have users directed to my own yet to be developed form, though this one may work if that's not possible.

I have a few, at first glance minor comments, which are not meant to be critical.  I haven't seen this before, and truly am trying to add what I hope is considered, good feedback. 

Can the text boxes for entering any information have the ability to be expanded, just as it's set up now for Notes?

Having the option to add a new meeting is great, though you'd have to get here by selecting a meeting and clicking or pressing "Request a Change" from opening the detail on any meeting. I suppose we can give instructions somewhere on the website.

When I selected the New option, I noticed both open and closed were checked, but a meeting is either open or it's closed -- not both. Ideally closed or open would be a radio button and not a check box, with the default being open. Same should apply to men or women with no default.

I am not sure what sub-region or sub-district is for.

Expanding to show Additional Group Information is great, but under Contact info, I was under the impression that info (name, phone & email) is always hidden from public view, something only the admins could see. I've been using it to keep track of who submitted changes. So not sure why we'd be asking for a non-personally identifying email since that would be hidden anyway. 

I filled out the form and purposely left things off like my email or put text in the email that wasn't an email. Pressing submit asked for confirmation I wanted to send. It wasn't sent because it wasn't a properly done request, and it was only when I at least put in a good submitter email that I received the thank you confirmation. Perhaps you can have the form alert people they have missing or incorrect information? 

I also put gibberish in the URL field, and even though I had a good email to the submitter box, it went through. 

The two emails I received didn't give me enough information to change the meeting or add a new one. 

I'd be glad to provide additional feedback you if you're interested in hearing more comments or suggestions. None of this is meant to be a slight in the least. 

Again, I want to emphasize how grateful I am for this plugin and all the great work that Code for Recovery team is doing for the recovery community. Countless numbers of people have benefitted from this.

**Reply by kiyote33** [reply]
Thanks for the feedback @Grateful-Viv on the TSMLEF plugin. Some of your comments make a lot of sense. Here's my thoughts on them:

1. **Can the text boxes for entering any information have the ability to be expanded, just as it's set up now for Notes?** Seems like a very reasonable request. I don't know if we would want to do it for every textbox, but certainly those boxes which tend to get more than a small number of characters input would be good candidates.
2. **Having the option to add a new meeting is great, though you'd have to get here by selecting a meeting and clicking or pressing "Request a Change" from opening the detail on any meeting.** It seems logical to group the 3 types of meeting change requests (change, new, remove) together at the same place. As you suggest, site designers can choose to add helpful information on how to make changes elsewhere.
3. **When I selected the New option, I noticed both open and closed were checked, but a meeting is either open or it's closed -- not both. Ideally closed or open would be a radio button and not a check box, with the default being open. Same should apply to men or women with no default.** The Types array listing is simply imported as is and displayed, with all items in the array handled with a single checkbox list. I like your suggestion to break specific types like open/close and men/women out from that array and assign them to radio button controls for display purposes. So I would support making this happen.
4. **I am not sure what sub-region or sub-district is for.** These fields support the hierarchical relationship of items in a WordPress taxonomy. For example, a website has categories books, politics, and blogging in it. While category itself is a taxonomy the items inside it are called terms. So, we can use these sub fields to relate specific parent/child terms to each other: a district to its parent area, a neighborhood to it's parent city, a city to its parent state, etc..
5. **Contact info, I was under the impression that info (name, phone & email) is always hidden from public view, something only the admins could see. I've been using it to keep track of who submitted changes. So not sure why we'd be asking for a non-personally identifying email since that would be hidden anyway.**  Groups, districts, central offices, or areas may or may not use these contact fields as they see fit. For example, a central office may ask groups in their intergroup area to provide contact information in these fields so groups can be contacted directly on issues which might affect them. This gives the group the ability to ensure they have responsible individuals registered with their local intergroup. Please note, the Signature field email address allows for anyone to submit a change request. They may or may not be someone who is appointed by the group conscience to speak for the group. The thought behind this approach is that it provides a feed back mechanism for someone showing up for a scheduled meeting, but there is no meeting. Web administrators will always need to use good judgement in validating the meeting information change requests.  (ie.. most popular meetings don't suddenly just stop, so a remove request might be questioned, or the meeting address of a local church probably won't change).
6. **Perhaps you can have the form alert people they have missing or incorrect information?** Agreed...
7. **I also put gibberish in the URL field, and even though I had a good email to the submitter box,** Gibberish in / Gibberish out. I suppose we could at least ensure only a valid URL format is input.
8. **The two emails I received didn't give me enough information to change the meeting or add a new one.** This is concerning. What info was missing?
9. **I'd be glad to provide additional feedback you if you're interested in hearing more comments or suggestions.**  Please do. No software is ever perfect when first released, but positive feedback certainly helps to make it better. Thank you.

**Reply by Grateful-Viv** [reply]
Thanks for the detail response @kiyote33 

I'll look in more detail on the TSMLEF plugin as soon as I get some other projects finished on my end. Our site is undergoing some changes with respect to the design, and that's where my efforts are focused on right now.

The question 3 above about open/closed or men/women I would think would be more of how TSML includes these options in types. Adding two more categories, one being open/closed and the other men/women should most likely be done, though that could be complicating things and maybe too much to ask. 

But back to my original question, is there a way that I can use my own form that will open if people want to have their meeting listing changed?

**Reply by kiyote33** [reply]
TSML doesn't currently support linking the feedback change request button to a custom link. We can discuss this at our next by-weekly meeting.

**Comment by joshreisner** [comment]
i think you can do this because of some code @97eighty added a few months ago, he added a [FAQ item on how to do it](https://wordpress.org/plugins/12-step-meeting-list/#can%20i%20add%20a%20feedback_url%20to%20each%20meeting%20when%20using%20tsml%20ui%3F%20%3D%20add%20a%20url%20to%20your%20themes%20functions.php.).

**Reply by Grateful-Viv** [reply]
Adding the following to the theme’s functions.php looks like I’d have to do it for each meeting. 

`$tsml_feedback_url = “https://domain.com?meeting={{slug}}”;
$tsml_feedback_url = “https://domain.com?meeting={{id}}”;
$tsml_feedback_url = “mailto:office@domain.com?subject={{slug}}”;`

If I were to do my own form, is there a way I can pass the meeting slug to the form? How was it setup for SF/Marin when someone wants to update a meeting? I would probably be using Gravity Forms. 

What @kiyote33 developed was good, but I’d prefer something simpler. One tab for meeting update info, including a checkbox or something like that to remove the meeting if that’s what the submitter intends to happen. No tab to add a new meeting — I would make a form for that.

**Reply by joshreisner** [reply]
my understanding is you add one of those lines, depending on the URL pattern you want, and it should work for all meetings.

perhaps @97eighty could advise on above? i think they're using gravity forms in santa cruz. we're also using them in San Jose, but looks like neither of us have enabled the feedback button yet.

https://aasantacruz.org/meeting-change-request/
https://aasanjose.org/new-updated-meeting-form

**Reply by 97eighty** [reply]
@joshreisner is correct. You only need the single entry to work with all meetings. In Gravity Forms the {{slug}} parameter (subject) could be read to auto populate a form. We used this originally but now query the meeting list directly.

**Reply by 97eighty** [reply]
@Grateful-Viv I just turned the feature back on in Santa Cruz. My functions.php contains this setting: 

$tsml_feedback_url = "https://aasantacruz.org/meeting-change-request/?meeting={{slug}}";

When used it passes meeting={{slug}} to the change request form. Gravity Forms allows you to pre-populate files with parameters. In this case I populate the hidden field "meeting slug" with the "meeting" parameter. The user doesn't see the field but it is stored in the database. You can view the source of the form and see the hidden field and its value.

**Reply by Grateful-Viv** [reply]
@97eighty 

You only put the one line in your functions.php about meeting slug? Not the others?

The example on your site opens the meeting form and I see the URL contain the meeting slug (for example, https://aasantacruz.org/meeting-change-request/?meeting=blf-womens-hour-of-power), but it doesn't appear any the current information is passed to the form when it's open. Or am I missing something?

**Reply by mikec-ncsd** [reply]
> i think you can do this because of some code @97eighty added a few months ago, he added a [FAQ item on how to do it](https://wordpress.org/plugins/12-step-meeting-list/#can%20i%20add%20a%20feedback_url%20to%20each%20meeting%20when%20using%20tsml%20ui%3F%20%3D%20add%20a%20url%20to%20your%20themes%20functions.php.).

Add a URL to your theme’s functions.php.

$tsml_feedback_url = '/feedback';

I added this to our website and see no change in behavior in either TSML UI or Legacy UI.

Am I missing something?

**Reply by joshreisner** [reply]
@mikec-ncsd you have to also edit a meeting. that refreshes the data

**Comment by 97eighty** [comment, marked as answer]
@Grateful-Viv Yes, Only a single line. The docs have several examples of the types of supported variables. 

Our Gravity Form contains a hidden field named "Meeting ID". This field type is hidden from view to the user. The field accepts a parameter named "meeting". When the "meeting" parameter is passed to the form the value is stored in the hidden field. In your example the value of "blf-womens-hour-of-power" is inserted into the "Meeting ID" field. When a user submits the form the "Meeting ID" field is submitted along with the other forms.

To see this without out hidden fields you can try this link.
https://aasantacruz.org/test/?meeting=blf-hour-of-power&ID=999

Several of the other popular form plugins support parameters too. 

Hope this helps,
Mark

**Reply by Grateful-Viv** [reply]
Thank you, Mark. I’ve used Gravity Forms before though not for this site, so I’ll think I’ll stick with it. I wasn’t crazy about their new design they rolled out last year (don’t use Gutenberg or block editors either), but I think it’s one of the best forms app around, even at their annual price. 

It may be a few weeks out before I get to this. I’m the only one providing support and there’s a few other priorities to tackle first. 

I may have some follow up questions, and if you don’t mind sharing the form you just referenced, that would be most appreciated.

**Comment by Grateful-Viv** [comment]
@97eighty, I wonder if you’d be willing to share with me a copy of your form and the snippet that you put into your functions.PHP. I know there is an export form feature in gravity forms. I would so appreciate that.

**Reply by 97eighty** [reply]
@Grateful-Viv, 

I can't export the form we use but have included a sample one the should work. Our form has too many custom mods and would require me to paste a bunch of code which may make it too complex. The setup is fairly simple. 

1. Create a form and under the advanced setting select "Allow field to populate dynamically". Create a parameter called "slug" or "id". The picture of this field is below:

<img width="386" alt="Screen Shot 2022-10-27 at 19 56 49" src="https://user-images.githubusercontent.com/72317744/198485406-33dfe930-2d96-4bff-bd8d-c106b5d915ed.png">

2. Create a blank page and embed the Gravity Form into the page. Note the name of the page permalink i.e. "change-form"

3. Insert $tsml_feedback_url = "https://yoursite.org/change-form/?slug={{slug}}" into your functions.php file. Change the URL.

4. Now, when the you the TSML-UI meeting change button url will load this URL which contains a URL parameter. In this example it is "slug". Gravity Forms will see this parameter and insert in into the field you allowed to populate dynamically. 

I have attached a sample form which should work with the URL I outlined above. This form contains two fields accepting the ID and SLUG parameters. You'll need to remove the .txt extension and leave the .json extension to import the form.

[gravityforms-export-2022-10-28.json.txt](https://github.com/code4recovery/12-step-meeting-list/files/9884786/gravityforms-export-2022-10-28.json.txt)

I'm not the best technical writer so let me know if anything is not clear. Happy to jump in and set it up for you.

**Reply by Grateful-Viv** [reply]
Mark

Trying to follow all this. I am sure it's simpler by seeing it. So let me see if I have some of this correct.

Each meeting has a slug and an ID through WordPress. Is that what you're referencing passing so that fields such as meeting name, day, time, etc. are auto-populated in the form? Do I get to pick which fields are auto-populated? 
 
For example, for this meeting: https://sussexaa.org/meetings/thursday-night-speakers/  the slug is **thursday-night-speakers** and the post ID is **6843**. 

So I make a GF, and make sure it includes hidden fields for ID and Slug with **Allow field to populate dynamically** checked. Create a page to embed the form. Let's say I call it Meeting-Change.

Then add **$tsml_feedback_url = "https://yoursite.org/change-form/?slug={{slug}}"** to functions.php and that's where it gets confusing since over [here ](https://wordpress.org/plugins/12-step-meeting-list/#can%20i%20add%20a%20feedback_url%20to%20each%20meeting%20when%20using%20tsml%20ui%3F%20%3D%20add%20a%20url%20to%20your%20themes%20functions.php), it says to enter the following three lines in my functions.php

**$tsml_feedback_url = “https://domain.org?meeting={{slug}}”;
$tsml_feedback_url = “https://domain.org?meeting={{id}}”;
$tsml_feedback_url = “mailto:office@domain.org?subject={{slug}}”;**

And now I am lost.

I appreciate your willingness to help with this. I know it goes beyond the scope for supporting this plugin.  Thank you.

**Reply by 97eighty** [reply]
@Grateful-Viv You have it. Perhaps, the docs need a little clarification. You only need one $tsml_feedback_url. The docs include those three examples. The {{ID}} and {{SLUG}} tags will be replaced by their true values. You can use any number of them.

**Reply by 97eighty** [reply]
@Grateful-Viv And you could use them together. i.e. https://domain.com/change-form/?id={{id}}&slug={{slug}}

**Reply by Grateful-Viv** [reply]
Dumb question perhaps… how do I tell it which fields to auto populate on the form?

**Comment by 97eighty** [comment]
@Grateful-Viv Not at all. 

Gravity Forms handles this automatically. When you mark a field "Allow field to populate dynamically", and include a parameter name (i.e. meeting), Gravity Forms looks for this parameter and its value when the form loads.

<img width="338" alt="Screen Shot 2022-11-01 at 17 30 05" src="https://user-images.githubusercontent.com/72317744/199366882-8d431302-53db-423d-9cff-9053058ba2c9.png">

In the above example my URL string would look something like: https://domain.com/change-form/?meeting={{slug}}. When parsed, WordPress creates this string: https://domain.com/change-form/?meeting=early-risers and populates the meeting change button.

When the form is loaded, Gravity Forms will see this parameter and populate the field with its value. In this case with "early-risers".

hth.

**Reply by Grateful-Viv** [reply]
Sorry to bring this up again. Finally got around to putting GF on a sandbox site. I added the following to our functions.php: $tsml_feedback_url = "https://sandbox.sussexaa.org/meeting-change/?meeting={{slug}}";

But when selecting Update this Listing from the meeting, it's still trying to do the email. And when opening the page with the form by putting in the slug (for example: https://sandbox.sussexaa.org/meeting-change/?meeting=all-are-welcome), it's not auto populating the fields such as meeting name.

I know it must be something obvious that I am missing. Appreciate your help with this.

**Reply by 97eighty** [reply]
@Grateful-Viv, I can jump in there are review if you like. DM me an access account.

**Reply by Grateful-Viv** [reply]
Not sure how to DM you on this platform. Do you have an email address I can reach out to you directly?

**Reply by 97eighty** [reply]
devs@code4recovery.org should arrive in my mailbox.

**Reply by Grateful-Viv** [reply]
Thanks. Sent you over an email from my intergroup account.

**Reply by 97eighty** [reply]
@Grateful-Viv, you're all set. I tweaked your $tsml_feedback_url variable to match the parameters you created in your form. The [Update Meeting Info] button now points to your form, and the values of the ID and SLUG are passed to the form. The cache had references to the MAILTO: links, so I refreshed a single meeting to force the meeting cache to be refreshed.

**Reply by Grateful-Viv** [reply]
> I tweaked your $tsml_feedback_url variable to match the parameters you created in your form

Thanks, @97eighty. Hopefully, I can recreate this on the production site.

But now how do I get the fields such as meeting name, meeting day, time, meeting types, etc. to auto-populate?

**Reply by Grateful-Viv** [reply]
How can I setup the form to auto populate fields such as meeting name, day, time, location, address, etc when it opens? That was always my intention of how it would work. 

In TSML Legacy, the link to update let the submitter put in a message relevant to the meeting and the email generated back to the admin displayed that message along with information about the meeting that included a link back to the meeting as depicted below.

![Meeting Feedback TSML(1)](https://user-images.githubusercontent.com/62247421/209561000-b746bd69-16f5-4e48-8f76-3fba794ee50c.png)

I realize TSML Legacy and TSML UI are different, and that's why I am trying to use a customized form. Is this possible to do?

**Reply by 97eighty** [reply]
This additional functionality is not native to TSML. Santa Cruz uses custom code and an add-on called Populate Anything by GravityWiz.

**Reply by Grateful-Viv** [reply]
Oh that’s disappointing. I thought this is how it could work, but it seems I didn’t understand correctly.

**Reply by Grateful-Viv** [reply]
> This additional functionality is not native to TSML. Santa Cruz uses custom code and an add-on called Populate Anything by GravityWiz.

I am still trying to make this work and was hoping at least meeting name could be auto-populated in addition to slug and ID without using the Populate Anything plugin.

The default meeting feedback link does generate an email with the subject containing the meeting name. Is this a big deal to do?

**Comment by joshreisner** [comment]
hey, just sort of catching up on this issue 

looks like we have an open TSML issue related to this: https://github.com/code4recovery/12-step-meeting-list/issues/849 - this proceeds along how Viv is going by adding variables to the `feedback_url` which get saved into the JSON feed. San Francisco uses this technique to send its users to a custom form - (they are not using TSML, but this would allow TSML to generate similar custom feedback_urls)

looks like viv's site is capturing ID and slug successfully. https://sandbox.sussexaa.org/meeting-change/?id=8026&slug=new-beginnings-step - @Grateful-Viv you would want to add more fields to your form now? are you able to name them anything you want? 

meanwhile @97eighty i am curious - looks like santa cruz is kinda still in progress? if i click to edit Nifty Ladies on the Park: https://aasantacruz.org/meeting-change-request/?meeting=nifty-ladies-on-park-group&title={{title}}&dow={{dow}} - looks like the variables aren't all there yet but its also using some other method of getting the right meeting name? 

so to move this forward do we need to with #849 ?  i wonder if we could do a shortcut and append all meeting variables to the URL? this would mean less setup in `functions.php` but would require specific field names

another approach would be to move forward with the [issue](https://github.com/code4recovery/tsml-ui/issues/252) and [PR](https://github.com/code4recovery/tsml-ui/pull/256) i started in TSML UI to replicate the behavior of the Enhanced Feedback plugin in TSML UI. here's an example page that doesn't do anything, but you can see what i might look like: https://deploy-preview-256--tsml-ui.netlify.app/tests/aasanjose.html?meeting=1010-monday

i'm not so sure if this is a great idea. the biggest hurdle to the TSML UI feedback form is that because it's a front-end system, it can't send emails on its own and needs help. it could get that from WordPress, but that does add some complexity and coordination. we could add support for a 3rd party service that people need to sign up for such as [EmailJS](https://www.emailjs.com/), but they are a for-profit company with sending limits. recently we also discussed developing a third party solution such as `feedback.code4recovery.org` - or creating another type of service.

also i really don't like the "new meeting" option - it seems bizarre to instruct users to request a new meeting by going to the bottom of an existing meeting page and requesting it there.

a lot to consider, wdyt?

**Reply by Grateful-Viv** [reply]
> looks like viv's site is capturing ID and slug successfully. https://sandbox.sussexaa.org/meeting-change/?id=8026&slug=new-beginnings-step - @Grateful-Viv you would want to add more fields to your form now? are you able to name them anything you want?

@joshreisner ideally everything on the meeting page, but if not possible, at least the meeting name. Then I’d leave them with the option to write up what change is requested in a paragraph field.

**Reply by Grateful-Viv** [reply]
@joshreisner I know you said you'd think about whether this was possible. I am wondering in the meantime if it's possible to get in addition to the meeting slug and ID, which already have, to have the meeting name ported over? 

The meeting slug is only useful to the admin, while the meeting name is understandable for the meeting change requestor.

Examples

bethany-12-12 is the slug, Bethany Beach Step is the meeting name.
wednesday-seaford-step is the slug, The Nanticoke Group of AA is the name
ocean-view-young-people is the slug, Uncut Group is the name

**Reply by joshreisner** [reply]
hey @Grateful-Viv we actually shipped an update to this in TSML recently that should enable all variables from the meeting to be able to be carried over -- what is your current code that's added to your functions.php? i can tell you how to expand it to include more variables

**Reply by Grateful-Viv** [reply]
@joshreisner, that is awesome. `$tsml_feedback_url = "https://sandbox.sussexaa.org/meeting-change/?meeting={{slug}}";`

Once I update the functions.php, do I need to know the exact parameter name? Currently have a field with the parameter name "id", and another parameter name "slug". I believe it doesn't matter what field label I give it, only what the parameter name is. 

Thank you.

**Reply by joshreisner** [reply]
oh great, so this is what it might look like with every single parameter enabled, i would try this:

```
$tsml_feedback_url = '/meeting-change/?meeting={{slug}}&id={{id}}&name={{name}}&day={{day}}&time={{time}}&end_time={{end_time}}&types={{types}}&notes={{notes}}&conference_url={{conference_url}}&conference_url_notes={{conference_url_notes}}&conference_phone={{conference_phone}}&conference_phone_notes={{conference_phone_notes}}&location={{location}}&formatted_address={{formatted_address}}&region={{region}}&location_notes={{location_notes}}&group={{group}}&group_notes={{group_notes}}&district={{district}}&website={{website}}&email={{email}}&phone={{phone}}&venmo={{venmo}}&square={{square}}&paypal={{paypal}}';
```

i'm guessing you're right you will need the form field name to match what is in the URL, see how our parameter is called `{{slug}}` but in the URL it's `meeting` -- you can change the values outside the braces as you need to, but for example the next step is probably to add a field for "Meeting Name" and make sure it's called `name` in the forms plugin

let me know if you run into problems, we are going to try to set this up on our demo site soon as well

**Reply by joshreisner** [reply]
playing around with this on our demo site, i am noticing a few problems, i think maybe because i used a few [wordpress reserved terms](https://codex.wordpress.org/Reserved_Terms) like `name`, `day`, and `types` -- probably best to avoid them for now.

```
$tsml_feedback_url = '/meeting-change/?meeting={{slug}}&id={{id}}&meetingName={{name}}&meetingDay={{day}}&time={{time}}&end_time={{end_time}}&meetingTypes={{types}}&notes={{notes}}&conference_url={{conference_url}}&conference_url_notes={{conference_url_notes}}&conference_phone={{conference_phone}}&conference_phone_notes={{conference_phone_notes}}&location={{location}}&formatted_address={{formatted_address}}&region={{region}}&location_notes={{location_notes}}&group={{group}}&group_notes={{group_notes}}&district={{district}}&website={{website}}&email={{email}}&phone={{phone}}&venmo={{venmo}}&square={{square}}&paypal={{paypal}}';
```

**Reply by Grateful-Viv** [reply]
@joshreisner I'm going to wait a short bit as I'm about to have a medical procedure. Hopefully, this will be sorted out meanwhile. At minimum, I'd like to have the meeting name which currently is {{name}} but it's a WordPress reserved word? 

Thank you so much for working on this.

**Reply by joshreisner** [reply]
i adjusted the query string so that the key is `meetingName` -- this is what you'd want to have your field look for for its default value

inside the braces we call it `name` because that's what TSML knows it as

eg https://code4recovery.org/feedback?meetingName=12+Keys+to+Freedom

**Reply by Grateful-Viv** [reply]
I used this in the functions.php. But am not able to get it to pick it up. It's still looking at the previous snippet. I know @97eighty said he had to tweak something to get it to work, but I don't know what that was. Any idea?

>@Grateful-Viv, you're all set. I tweaked your $tsml_feedback_url variable to match the parameters you created in your form. The [Update Meeting Info] button now points to your form, and the values of the ID and SLUG are passed to the form. The cache had references to the MAILTO: links, so I refreshed a single meeting to force the meeting cache to be refreshed.

`$tsml_feedback_url = '/meeting-change/?meeting={{slug}}&id={{id}}&meetingName={{name}}&meetingDay={{day}}&time={{time}}&end_time={{end_time}}&meetingTypes={{types}}&notes={{notes}}&conference_url={{conference_url}}&conference_url_notes={{conference_url_notes}}&conference_phone={{conference_phone}}&conference_phone_notes={{conference_phone_notes}}&location={{location}}&formatted_address={{formatted_address}}&region={{region}}&location_notes={{location_notes}}&group={{group}}&group_notes={{group_notes}}&district={{district}}&website={{website}}&email={{email}}&phone={{phone}}&venmo={{venmo}}&square={{square}}&paypal={{paypal}}';`

**Reply by joshreisner** [reply]
you will need to save changes on a meeting to get the URLs to update

**Reply by Grateful-Viv** [reply]
> you will need to save changes on a meeting to get the URLs to update

I have done that several times. Then purged cache. Must have done something obviously wrong that I cannot figure out.

**Reply by joshreisner** [reply]
hmm! maybe double check that you don't have the old code in addition to the new one? and/or double-check that the functions change and the refresh are both happening on the same staging server?

definitely seems to all be working properly on our demo site: https://code4recovery.org/tsml-ui

**Reply by Grateful-Viv** [reply]
I got it to work. Thank you!

**Reply by Grateful-Viv** [reply]
Started to slowly work on this. Want to verify that every time I change the form that I have to go back to any meeting an update it so it can get a refresh? Otherwise will get form not found. Is that correct?

Also address fields, phone numbers, etc should really be a text field in Gravity Forms?

**Reply by joshreisner** [reply]
i don't have gravity forms, but i don't think so. i think you'll only need to update a meeting if you change your `$tsml_feedback_url ` in functions.php

when you make a change to the form i would think the URL would stay the same

**Reply by Grateful-Viv** [reply]
I’m finding I need to do this. It’s no big deal in the long run. Thanks @joshreisner

**Reply by joshreisner** [reply]
hm, maybe it's a non-tsml caching issue?

**Reply by Grateful-Viv** [reply]
Cache issue. Thank you again.

**Reply by Grateful-Viv** [reply]
Is there a way to capture the attendance options? Online, TC, In-person, hybrid, etc?

**Reply by joshreisner** [reply]
not as such. `TC` will be in the query string under `meetingTypes`. the rest we calculate based on logic: https://github.com/code4recovery/tsml-ui#how-are-metatypes-like-active-and-online-calculated

i can't test this so i'm just freestyling, but i think a PHP script might look something like:

```php
// online if url or phone is not empty
$isOnline = !empty(rgget( 'conference_url' )) || !empty(rgget( 'conference_phone' ));

// location temporarily closed?
$meetingTypes = explode(',', rgget( 'meetingTypes' ));
$isTempClosed = in_array('TC', $meetingTypes);

// if address has fewer than 3 commas, it's approximate (normally we would geocode)
$addressParts = explode(',', rgget( 'formatted_address' ));
$isApproximate = count($addressParts) < 4; // count commas

$isInPerson = !$isTempClosed && !$isApproximate;

if ($isOnline && $isInPerson) {
  $attendance = 'Hybrid';
} else if ($isOnline) {
  $attendance = 'Online';
} else if ($isInPerson) {
  $attendance = 'In person';
} else {
  $attendance = 'Inactive';
}
```

Now you can use the `$attendance` variable to output the meeting attendance type, eg

```php
echo "Attendance is: $attendance";
```
